### DISCUSSION TITLE
How do I get FrankenPHP's Caddy running from Octane? (connection refused when using --caddyfile)

### DISCUSSION THREAD

**Discussion by nikspyratos** [php/frankenphp#673] [Q&A] [answered]
👋 Heyo.

Been playing around with FrankenPHP on and off to get it working and haven't had much success with the various modes. I'm on latest macOS with PHP 8.3 (Herd/Valet are turned off while doing this).

Based on [this convo](https://twitter.com/NikSpyratos/status/1765262522922291408) and having the `--caddyfile` option in Octane, my impression is that running `php artisan octane:start --caddyfile=MyCaddyfile` should start both a Caddy server and the Octane processes, right?

Octane starts, but every time I try to load the page, I get a 502 response and all I see in the octane output is:
```
ERROR  dial tcp 127.0.0.1:8000: connect: connection refused
```

If I run Caddy separately from Octane via `frankenphp run --caddyfile=MyCaddyfile` (I have a separate Caddyfile for local dev), everything works.

Am I misunderstanding what `--caddyfile` is supposed to do in the Octane command? Or is there another way to run all this from one command instead of two?

**Comment by dunglas** [comment]
`--caddyfile` is an advanced option to use a custom Caddy config. You don't need it by default. `php artisan octane:start --server=frankenphp` should work for most use cases.

**Reply by nikspyratos** [reply]
Thanks.

I think my confusion lies in that just running `artisan octane:start --server=frankenphp` would _not_ start the Caddy server as well right? In which case it's unclear to me what `--caddyfile` does at all when run in Octane. AFAIK Octane is intended to still run behind a webserver, not just straight out to the internet.

If it _does_ start the webserver as well as Octane, I have no idea how to debug the `connection refused` part.

**Reply by dunglas** [reply, marked as answer]
FrankenPHP **is** a build of the Caddy webserver, with an extra PHP module (similar to Apache + mod_php).

Octane + FrankenPHP can be exposed directly on the internet (but not Octane + Swoole or RoadRunner).

**Reply by nikspyratos** [reply]
I see thanks! That's what I thought but was hard to determine based on the error I was getting.

Guess I'll try with Docker as well and see how it goes.

If nothing else hope this question helps others searching for the same too :) 

EDIT: Nvm, figured out the error. Bit of PEBCAK and false memory. All working now!

**Reply by fpolli** [reply]
> FrankenPHP **is** a build of the Caddy webserver, with an extra PHP module (similar to Apache + mod_php).
> Octane + FrankenPHP can be exposed directly on the internet (but not Octane + Swoole or RoadRunner).

Hi, building off this, the Octane docs suggest:

`In production environments, you should serve your Octane application behind a traditional web server such as Nginx or Apache. Doing so will allow the web server to serve your static assets such as images and stylesheets, as well as manage your SSL certificate termination.`

This confuses me because Caddy is supposed to do all those things, so why add the Nginx layer? Doesn't that just slow everything down and take away a lot of what FrankenPHP gives us? If FrankenPHP is a build of Caddy we should be able to configure Caddy to do those things. What am I missing?

Thanks!

**Reply by nikspyratos** [reply]
@fpolli I think that's the Octane docs not accounting for using the Caddy portion of FrankenPHP, not specifically a FrankenPHP issue.

**Reply by fpolli** [reply]
Thank you for answering. That is what I thought but I wanted to confirm and resolve the contradiction. I am excited to get into FrankenPHP.
